Modules/Core/Common/src/itkMersenneTwisterRandomVariateGenerator.cxx

```diff
@@ -89,31 +89,28 @@ MersenneTwisterRandomVariateGenerator::MersenneTwisterRandomVariateGenerator()
 MersenneTwisterRandomVariateGenerator::~MersenneTwisterRandomVariateGenerator() = default;
 
 MersenneTwisterRandomVariateGenerator::IntegerType
-MersenneTwisterRandomVariateGenerator ::hash(time_t t, clock_t c)
+MersenneTwisterRandomVariateGenerator::hash(const time_t t, const clock_t c)
 {
   itkInitGlobalsMacro(PimplGlobals);
   // Get an IntegerType from t and c
   // Better than IntegerType(x) in case x is floating point in [0,1]
   // Based on code by Lawrence Kirby: fred at genesis dot demon dot co dot uk
 
-  IntegerType h1 = 0;
-  auto *      p = (unsigned char *)&t;
-
-  const auto sizeOfT = static_cast<unsigned int>(sizeof(t));
-  for (unsigned int i = 0; i < sizeOfT; ++i)
-  {
-    h1 *= UCHAR_MAX + 2U;
-    h1 += p[i];
-  }
-  IntegerType h2 = 0;
-  p = (unsigned char *)&c;
-
-  const auto sizeOfC = static_cast<unsigned int>(sizeof(c));
-  for (unsigned int j = 0; j < sizeOfC; ++j)
-  {
-    h2 *= UCHAR_MAX + 2U;
-    h2 += p[j];
-  }
+  const auto convert = [](const auto arg) {
+    IntegerType        h = 0;
+    const void * const voidPtr = &arg;
+    auto * const       unsignedCharPtr = static_cast<const unsigned char *>(voidPtr);
```

Comment:
Why not use reinterpret cast here?

Reply:
Let's document in the conversation here (and maybe in the code too) why it makes sense to do the pointer conversion via an intermediate `void *`.  Or if it is reasonable to avoid `void *`, as the original code does, then let's do that instead, e.g., via the `reinterpret_cast`.

Reply:
>Why not use reinterpret cast here?

@dzenanz  Thanks for asking. I was told that the effect of reinterpret_cast is "implementation defined" (compiler dependent), and that it should be avoided whenever there is a reasonably safe alternative. Specifically, [C++ Standard Working Draft n4910](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/n4910.pdf) 7.6.1.10 Reinterpret cast [expr.reinterpret.cast] says:

> A pointer can be explicitly converted to any integral type large enough to hold all values of its type. The
mapping function is implementation-defined.

Implicit conversion of an address to a `const void ` pointer is entirely safe, though. And a cast from such a pointer to `const unsigned char *` is OK as well.

Reply:
> Let's document in the conversation here (and maybe in the code too)

Thanks for you suggestion @Leengit I wouldn't mind a line of comment here, but then, I would rather not see such a comment copy-and-pasted throughout the code base, whenever an address is converted to an `unsigned char *`. So do you think we need an extra helper function template, `itk::CastToUnsignedCharPtr(T*) -> unsigned char* `, if such conversion would take place more often ?

Reply:
My reading of http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/n4910.pdf 7.6.1.10, bullet 7 is that skipping the cast to `void *` is required to be equivalent to having used `void *` as an intermediate.

Either way, unfortunately C++ doesn't seem to guarantee that we can do what we want to do here via pointer manipulation.  (I think we should do it anyway.  An alternative is to look at `t & 0xFF`, `t >> 8 & 0xFF`, etc.)

Reply:
I suppose we could side-step the issue with something like
```cpp
  const auto convert = [](auto arg) {
    IntegerType h{ 0 };
    for (size_t i{ 0 }; i < sizeof(arg); ++i)
    {
      h *= UCHAR_MAX + 2U;
      h += (arg & 0xFF);
      arg >>= 8;
    }
    return h;
  };
```

Reply:
@dzenanz @Leengit You converted me to use reinterpret_cast  😸  With [this force-push](https://github.com/InsightSoftwareConsortium/ITK/compare/8583dba865362e4c30a169848ed19329f67b8945..f68a49aa157ee3f8324da429f7ad076edfef8037)

Reply:
Looks like this code by Lawrence Kirby is quite popular  😄 I encountered similar lines of code on various sites, including: 

  http://beagle.gel.ulaval.ca/refmanual/2.2.0/RandomizerMersenne_8cpp-source.html 
  https://groups.google.com/g/comp.lang.c/c/smOwxlZVbeA (post by Ben Pfaff)
  https://github.com/spcl/foMPI-NA/blob/master/MersenneTwister.h#L319-L340

Reply:
> @Leengit Nice suggestion, `(arg & 0xFF)`, but I'm not entirely sure if it would work for any `clock_t` or `time_t` implementation. You?

If `clock_t` or `time_t` could ever be other than a `signed` or `unsigned` `int` (of any size including `char`), we'd be in trouble.  I can't say that I have surveyed every system, but I think there is a decent probability that this  approach would actually work.  (But your solution is  good, so there is no need to pursue this further.)